PKCS11: Test Cases for Object APIs - Finding Objects

ruchi393 opened the pull request:
Add 1011 for -
C_FindObjectsInit()
C_FindObjects()
C_FindObjectsFinal()

Signed-off-by: Ruchika Gupta <ruchika.gupta@linaro.org>

<!--
    If you are new to submitting pull requests to OP-TEE, then please have a
    look at the list below and tick them off before submitting the pull request.

    1. Read our contribution guidelines:
         https://optee.readthedocs.io/general/contribute.html

    2. Read the contribution section in Notice.md and pay extra attention to the
       "Developer Certificate of Origin" in the contribution guidelines.

    3. You should run checkpatch preferably before submitting the pull request.

    4. When everything has been reviewed, you will need to squash, rebase and
       add tags like `Reviewed-by`, `Acked-by`, `Tested-by` etc. More details
       about this can also be found on the link provided above.

    NOTE: This comment will not be shown in the pull request, so no harm keeping
    it, but feel free to remove it if you like.
-->

ruchi393 commented:
This is dependent on following PR's
https://github.com/OP-TEE/optee_os/pull/4286
https://github.com/OP-TEE/optee_os/pull/4288
https://github.com/OP-TEE/optee_client/pull/244

vesajaaskelainen commented on host/xtest/pkcs11_1000.c:
```diff
@@ -2114,3 +2114,417 @@ static void xtest_pkcs11_test_1010(ADBG_Case_t *c)
+	};
+	CK_ATTRIBUTE find_session_template[] = {
+		{ CKA_LABEL, (CK_UTF8CHAR_PTR)label, strlen(label) },
+		{ CKA_TOKEN, &(CK_BBOOL){CK_FALSE}, sizeof(CK_BBOOL) },
```
To exercise the serializer also a bit it is a good idea to have CK_ULONG attributes (as they have special handling) and CK_BBOOL attributes (they are only 1 byte) in between there.

etienne-lms replied:
Maybe add specific extra tests in a later change?

vesajaaskelainen replied:
I am OK with that. When we have set and copy support it is much easier to make all kind of tests for attributes.

ruchi393 replied:
Thanks @etienne-lms , @vesajaaskelainen . I will take a note and add a case in it with set and copy support.

etienne-lms commented on host/xtest/pkcs11_1000.c:
```diff
@@ -2114,3 +2114,417 @@ static void xtest_pkcs11_test_1010(ADBG_Case_t *c)
+	ADBG_EXPECT_CK_OK(c, rv);
+
+	/*
+	 * Intentionlly do not finalize the active object search. It should
```
s/Intentionlly/Intentionally/

ruchi393 replied:
Done

etienne-lms reviewed (commented):
~~`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>`~~
(edited: removed this content changes)

etienne-lms commented on host/xtest/pkcs11_1000.c:
```diff
@@ -2114,3 +2114,417 @@ static void xtest_pkcs11_test_1010(ADBG_Case_t *c)
+	if (!ADBG_EXPECT_CK_RESULT(c, CKR_USER_NOT_LOGGED_IN, rv))
+		goto out;
+
+	object_id--;
```
should remove this line and `++` incremention from line 2282.

ruchi393 replied:
Done

etienne-lms commented on host/xtest/pkcs11_1000.c:
```diff
@@ -2312,6 +2314,17 @@ static void xtest_pkcs11_test_1011(ADBG_Case_t *c)
 	if (!ADBG_EXPECT_CK_OK(c, rv))
 		goto out;
 
+	/*
```
You forgot to move `Do_ADBG_BeginSubCase()`/`Do_ADBG_EndSubCase()`.

ruchi393 replied:
I added this as a check to ensure all handles are valid. Its not a new Subcase.

etienne-lms replied:
Ok, then, should remove sub test 9 that is now empty.

ruchi393 replied:
Sorry for that. Already added a commit for same.

etienne-lms commented on host/xtest/pkcs11_1000.c:
```diff
@@ -2114,3 +2114,417 @@ static void xtest_pkcs11_test_1010(ADBG_Case_t *c)
+	CK_SESSION_HANDLE session = CK_INVALID_HANDLE;
+	CK_FLAGS session_flags = CKF_SERIAL_SESSION | CKF_RW_SESSION;
+	CK_OBJECT_HANDLE obj_hdl[10];
+	CK_OBJECT_HANDLE obj_found[10];
```
not addressed

ruchi393 replied:
Sorry, missed this earlier. Have fixed it now.

etienne-lms commented on host/xtest/pkcs11_1000.c:
```diff
@@ -2114,3 +2114,441 @@ static void xtest_pkcs11_test_1010(ADBG_Case_t *c)
+	 * Check if object handles returned when creating objects with this
+	 * session are still valid
+	 */
+	for (uint32_t i = 0; i < object_id; i++) {
```
Prefer `uint32_t i = 0;` definition at function entry.
See also the other `for()` loop.

ruchi393 replied:
Done

etienne-lms commented on host/xtest/pkcs11_1000.c:
```diff
@@ -2114,3 +2114,441 @@ static void xtest_pkcs11_test_1010(ADBG_Case_t *c)
+	Do_ADBG_EndSubCase(c, NULL);
+
+	/*
+	 * Sub test 8: Close session and open new session , find objects
```
space char before coma.

ruchi393 replied:
Fixed

etienne-lms reviewed (commented):
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`

ruchi393 commented:
> `Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`

Thanks @etienne-lms . I will squash commits and apply tags once the optee_os and optee_client patches are accepted.

etienne-lms commented on host/xtest/pkcs11_1000.c:
```diff
@@ -2114,3 +2114,442 @@ static void xtest_pkcs11_test_1010(ADBG_Case_t *c)
+
+	rv = test_find_objects(c, session, find_template,
+			       ARRAY_SIZE(find_template),
+			       obj_found, ARRAY_SIZE(obj_found), 4);
```
These line could be change to test that providing no attribute return all visible objects:
```patch
-	rv = test_find_objects(c, session, find_template,
-			       ARRAY_SIZE(find_template),
+	rv = test_find_objects(c, session, NULL, 0,
 			       obj_found, ARRAY_SIZE(obj_found), 4);
```

ruchi393 replied:
@etienne-lms , saw your comment late. I just added another sub-case for this.

etienne-lms commented:
My `Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` still stands.

ruchi393 commented:
Squashed, rebased and tags applied. 

This PR has a dependency on 
https://github.com/OP-TEE/optee_os/pull/4288
https://github.com/OP-TEE/optee_client/pull/244

jforissier commented:
> Squashed, rebased and tags applied.
> 
> This PR has a dependency on
> [OP-TEE/optee_os#4288](https://github.com/OP-TEE/optee_os/pull/4288)
> [OP-TEE/optee_client#244](https://github.com/OP-TEE/optee_client/pull/244)

All, do you think it's OK for me to merge those PRs once they are properly reviewed, even though we're post 3.12.0-rc1? My opinion is: yes it is, since PKCS#11 is still experimental and incomplete so the risk of regression is quite minimal.

etienne-lms commented:
Hi Jérôme, I think it is safe too, for the reasons you gave.

etienne-lms commented:
... but of course let's wait OP-TEE/optee_os#4288 & OP-TEE/optee_client#244 have landed before merging this  optee_test P-R :)

vesajaaskelainen commented:
Did quick look and did not spot anything alarming. If you want to go ahead with merge feel free and no need to wait me for doing better review for test cases here.

ruchi393 commented:
Thanks @vesajaaskelainen .

@jforissier , we can probably merge the  OP-TEE/optee_os#4288 & OP-TEE/optee_client#244 now.  Both those PR's have R-by tags applied now.  That would also fix the  IBART errors in this PR and we can consider merging it.

jforissier commented:
@ruchi393 yes I agree, I have seen the updates and I'm just waiting for CI to finish before merging. Thanks.
